File: src/main/java/com/example/MicroInvestApp/service/learning/EducationalContentService.java
package com.example.MicroInvestApp.service.learning;

import com.example.MicroInvestApp.domain.enums.ContentCategory;
import com.example.MicroInvestApp.domain.enums.ContentType;
import com.example.MicroInvestApp.domain.enums.DifficultyLevel;
import com.example.MicroInvestApp.dto.edcontent.EducationalContentResponseDTO;
import com.example.MicroInvestApp.dto.edcontent.CreateEducationalContentRequestDTO;
import com.example.MicroInvestApp.dto.edcontent.UpdateEducationalContentRequestDTO;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public interface EducationalContentService {

    /**
     * Create new educational content
     * @param request the create request DTO
     * @return created educational content response DTO
     */
    EducationalContentResponseDTO createContent(CreateEducationalContentRequestDTO request);

    /**
     * Get educational content by ID
     * @param contentId the content ID
     * @return Optional educational content response DTO
     */
    Optional<EducationalContentResponseDTO> getContentById(Long contentId);

    /**
     * Get educational content by ID with user progress
     * @param contentId the content ID
     * @param userId the user ID
     * @return Optional educational content response DTO with user progress
     */
    Optional<EducationalContentResponseDTO> getContentByIdWithProgress(Long contentId, Long userId);

    /**
     * Update educational content
     * @param contentId the content ID
     * @param request the update request DTO
     * @return updated educational content response DTO
     */
    EducationalContentResponseDTO updateContent(Long contentId, UpdateEducationalContentRequestDTO request);

    /**
     * Delete educational content
     * @param contentId the content ID
     */
    void deleteContent(Long contentId);

    /**
     * Update content status (active/inactive)
     * @param contentId the content ID
     * @param isActive the active status
     * @return updated educational content response DTO
     */
    EducationalContentResponseDTO updateContentStatus(Long contentId, boolean isActive);

    /**
     * Get all active educational content
     * @return List of active educational content
     */
    List<EducationalContentResponseDTO> getAllActiveContent();

    /**
     * Get all active educational content with user progress
     * @param userId the user ID
     * @return List of active educational content with user progress
     */
    List<EducationalContentResponseDTO> getAllActiveContentWithProgress(Long userId);

    /**
     * Get educational content by category
     * @param category the content category
     * @return List of educational content
     */
    List<EducationalContentResponseDTO> getContentByCategory(ContentCategory category);

    /**
     * Get educational content by category with user progress
     * @param category the content category
     * @param userId the user ID
     * @return List of educational content with user progress
     */
    List<EducationalContentResponseDTO> getContentByCategoryWithProgress(ContentCategory category, Long userId);

    /**
     * Get educational content by difficulty level
     * @param difficultyLevel the difficulty level
     * @return List of educational content
     */
    List<EducationalContentResponseDTO> getContentByDifficultyLevel(DifficultyLevel difficultyLevel);

    /**
     * Get educational content by content type
     * @param contentType the content type
     * @return List of educational content
     */
    List<EducationalContentResponseDTO> getContentByType(ContentType contentType);

    /**
     * Get featured educational content
     * @return List of featured educational content
     */
    List<EducationalContentResponseDTO> getFeaturedContent();

    /**
     * Get featured educational content with user progress
     * @param userId the user ID
     * @return List of featured educational content with user progress
     */
    List<EducationalContentResponseDTO> getFeaturedContentWithProgress(Long userId);

    /**
     * Search educational content by title
     * @param title the title search term
     * @return List of matching educational content
     */
    List<EducationalContentResponseDTO> searchContentByTitle(String title);

    /**
     * Search educational content by author
     * @param author the author search term
     * @return List of matching educational content
     */
    List<EducationalContentResponseDTO> searchContentByAuthor(String author);

    /**
     * Get educational content by category and difficulty level
     * @param category the content category
     * @param difficultyLevel the difficulty level
     * @return List of educational content
     */
    List<EducationalContentResponseDTO> getContentByCategoryAndDifficulty(ContentCategory category, DifficultyLevel difficultyLevel);

    /**
     * Get educational content created after date
     * @param date the date
     * @return List of educational content
     */
    List<EducationalContentResponseDTO> getContentCreatedAfter(LocalDateTime date);

    /**
     * Get educational content with minimum rating
     * @param minRating the minimum rating
     * @return List of educational content
     */
    List<EducationalContentResponseDTO> getContentByMinimumRating(Double minRating);

    /**
     * Get most popular educational content
     * @param limit the maximum number of results
     * @return List of most popular educational content
     */
    List<EducationalContentResponseDTO> getMostPopularContent(int limit);

    /**
     * Get educational content by duration range
     * @param minDuration minimum duration in minutes
     * @param maxDuration maximum duration in minutes
     * @return List of educational content
     */
    List<EducationalContentResponseDTO> getContentByDurationRange(Integer minDuration, Integer maxDuration);

    /**
     * Search educational content by tags
     * @param tag the tag to search for
     * @return List of matching educational content
     */
    List<EducationalContentResponseDTO> searchContentByTag(String tag);

    /**
     * Increment view count for content
     * @param contentId the content ID
     */
    void incrementViewCount(Long contentId);

    /**
     * Update content rating
     * @param contentId the content ID
     * @param rating the new rating
     * @return updated educational content response DTO
     */
    EducationalContentResponseDTO updateContentRating(Long contentId, Double rating);

    /**
     * Get content statistics
     * @return Map containing various statistics
     */
    Map<String, Object> getContentStatistics();

    /**
     * Get content statistics by category
     * @return Map containing statistics by category
     */
    Map<ContentCategory, Long> getContentStatsByCategory();

    /**
     * Get total active content count
     * @return total count of active content
     */
    Long getTotalActiveContentCount();

    /**
     * Get average content rating
     * @return average rating across all content
     */
    Double getAverageContentRating();

    /**
     * Validate content exists and is active
     * @param contentId the content ID
     * @return true if content exists and is active
     */
    boolean validateContentExists(Long contentId);
}

File: frontend/src/services/educationalContentService.ts
// services/educationalContentService.ts
import type {
  EducationalContent,
  CreateEducationalContentRequest,
  UpdateEducationalContentRequest,
  ContentFilters,
  ContentStatistics,
  UserProgress
} from "../types/educationalContent";
import { 
  ContentCategory, 
  ContentType, 
  DifficultyLevel 
} from "../types/educationalContent";
import { apiClient } from "./api";

export const educationalContentService = {
  // Get all active educational content
  async getAllContent(userId?: number): Promise<EducationalContent[]> {
    const params = userId ? { userId } : {};
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content", { params });
  },

  // Get content by ID
  async getContentById(contentId: number, userId?: number): Promise<EducationalContent> {
    const url = userId 
      ? `/api/v1/educational-content/${contentId}/progress?userId=${userId}`
      : `/api/v1/educational-content/${contentId}`;
    return apiClient.get<EducationalContent>(url);
  },

  // Get featured content
  async getFeaturedContent(userId?: number): Promise<EducationalContent[]> {
    const params = userId ? { userId } : {};
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content/featured", { params });
  },

  // Get content by category
  async getContentByCategory(category: ContentCategory, userId?: number): Promise<EducationalContent[]> {
    const params = userId ? { userId } : {};
    return apiClient.get<EducationalContent[]>(`/api/v1/educational-content/category/${category}`, { params });
  },

  // Get content by difficulty level
  async getContentByDifficulty(difficulty: DifficultyLevel): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>(`/api/v1/educational-content/difficulty/${difficulty}`);
  },

  // Get content by type
  async getContentByType(type: ContentType): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>(`/api/v1/educational-content/type/${type}`);
  },

  // Search content by title
  async searchByTitle(title: string): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content/search/title", {
      params: { title }
    });
  },

  // Search content by author
  async searchByAuthor(author: string): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content/search/author", {
      params: { author }
    });
  },

  // Search content by tag
  async searchByTag(tag: string): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content/search/tag", {
      params: { tag }
    });
  },

  // Get content by duration range
  async getContentByDurationRange(minDuration: number, maxDuration: number): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content/duration-range", {
      params: { minDuration, maxDuration }
    });
  },

  // Get content by minimum rating
  async getContentByMinRating(minRating: number): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content/rating", {
      params: { minRating }
    });
  },

  // Get most popular content
  async getMostPopularContent(limit: number = 10): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content/popular", {
      params: { limit }
    });
  },

  // Get recent content
  async getRecentContent(date: string): Promise<EducationalContent[]> {
    return apiClient.get<EducationalContent[]>("/api/v1/educational-content/created-after", {
      params: { date }
    });
  },

  // Increment view count
  async incrementViewCount(contentId: number): Promise<void> {
    return apiClient.post(`/api/v1/educational-content/${contentId}/view`);
  },

  // Get content statistics (Admin only)
  async getContentStatistics(): Promise<ContentStatistics> {
    return apiClient.get<ContentStatistics>("/api/v1/educational-content/statistics");
  },

  // Validate content exists
  async validateContentExists(contentId: number): Promise<{ exists: boolean }> {
    return apiClient.get<{ exists: boolean }>(`/api/v1/educational-content/${contentId}/validate`);
  },

  // Admin functions
  async createContent(request: CreateEducationalContentRequest): Promise<EducationalContent> {
    return apiClient.post<EducationalContent>("/api/v1/educational-content", request);
  },

  async updateContent(contentId: number, request: UpdateEducationalContentRequest): Promise<EducationalContent> {
    return apiClient.put<EducationalContent>(`/api/v1/educational-content/${contentId}`, request);
  },

  async deleteContent(contentId: number): Promise<void> {
    return apiClient.delete(`/api/v1/educational-content/${contentId}`);
  },

  async updateContentStatus(contentId: number, isActive: boolean): Promise<EducationalContent> {
    return apiClient.patch<EducationalContent>(`/api/v1/educational-content/${contentId}/status`, null, {
      params: { isActive }
    });
  },

  // Advanced filtering
  async getFilteredContent(filters: ContentFilters, userId?: number): Promise<EducationalContent[]> {
    // Build query based on filters
    let endpoint = "/api/v1/educational-content";
    const params: any = {};

    if (userId) {
      params.userId = userId;
    }

    // Handle different filter types
    if (filters.category) {
      endpoint = `/api/v1/educational-content/category/${filters.category}`;
    } else if (filters.difficulty_level) {
      endpoint = `/api/v1/educational-content/difficulty/${filters.difficulty_level}`;
    } else if (filters.content_type) {
      endpoint = `/api/v1/educational-content/type/${filters.content_type}`;
    } else if (filters.is_featured) {
      endpoint = "/api/v1/educational-content/featured";
    } else if (filters.search_term) {
      endpoint = "/api/v1/educational-content/search/title";
      params.title = filters.search_term;
    } else if (filters.author) {
      endpoint = "/api/v1/educational-content/search/author";
      params.author = filters.author;
    } else if (filters.min_rating) {
      endpoint = "/api/v1/educational-content/rating";
      params.minRating = filters.min_rating;
    } else if (filters.min_duration && filters.max_duration) {
      endpoint = "/api/v1/educational-content/duration-range";
      params.minDuration = filters.min_duration;
      params.maxDuration = filters.max_duration;
    }

    return apiClient.get<EducationalContent[]>(endpoint, { params });
  },

  // Get all categories for filtering
  getCategories(): ContentCategory[] {
    return Object.values(ContentCategory);
  },

  // Get all difficulty levels
  getDifficultyLevels(): DifficultyLevel[] {
    return Object.values(DifficultyLevel);
  },

  // Get all content types
  getContentTypes(): ContentType[] {
    return Object.values(ContentType);
  }
};